### PR TITLE
Download all attachments of an email as zip

### PR DESCRIPTION
**Pull request by miaulalala**
To Do:

- [x] Integrate button on attachment modal
- [x] Wire up to backend
- [x] Tests

See #552

### CODE REVIEW

File: lib/Controller/MessagesController.php [hunk 1]
```diff
@@ -525,6 +526,43 @@ public function downloadAttachment(int $id,
 		);
 	}
 
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 * @TrapError
+	 *
+	 * @param int $id - the message id
+	 * @param string $attachmentId
```
Comment (ChristophWurst) [inline lib/Controller/MessagesController.php:535]: nope ;)

File: lib/Controller/MessagesController.php [hunk 2]
```diff
@@ -525,6 +526,43 @@ public function downloadAttachment(int $id,
 		);
 	}
 
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 * @TrapError
+	 *
+	 * @param int $id - the message id
```
Comment (ChristophWurst) [inline lib/Controller/MessagesController.php:534]:
```suggestion
	 * @param int $id the message id
```

File: src/components/MessageAttachments.vue [hunk 3]
```diff
@@ -67,12 +79,18 @@ export default {
 	data() {
 		return {
 			savingToCloud: false,
+			zipping: false,
```
Comment (ChristophWurst) [inline src/components/MessageAttachments.vue:82]: unused?

Comment (miaulalala) [reply src/components/MessageAttachments.vue:82]: Ah yes, thanks.

File: lib/Controller/MessagesController.php [hunk 4]
```diff
@@ -525,6 +526,43 @@ public function downloadAttachment(int $id,
...
+		$attachments = $this->mailManager->getMailAttachments($account, $mailbox, $message);
+		$zip = new ZipResponse($this->request, 'attachments');
+
+		foreach ($attachments as $attachment) {
+			$fileName = $attachment['name'];
+			$fh = fopen("php://temp", 'r+');
+			fputs($fh, $attachment['content']);
+			$size = (int)$attachment['size'];
+			rewind($fh);
+			$zip->addResource($fh, $fileName, $size);
```
Comment (st3iny) [inline lib/Controller/MessagesController.php:561]:
```suggestion
			$fileName = $attachment['name'];
			$fh = fopen("php://temp", 'r+');
			fputs($fh, $attachment['content']);
			$size = (int)$attachment['size'];
			rewind($fh);
			$zip->addResource($fh, $fileName, $size);
			fclose($fh);
```

Does php automatically close files that go out of scope? If not it should be closed manually here.

**Review by st3iny — APPROVED** [review]
Just tested this branch and it works well. Thanks for your contribution :)

File: lib/Controller/MessagesController.php (hunk 4 above)
Comment (ChristophWurst) [reply lib/Controller/MessagesController.php:561]: After having a deeper look I think it's not quite right to close the stream here already. The reason is that `addResource` won't consume the stream right away but only when the response data is rendered and streamed. So if we close this here, the mem stream won't be available for the zip data later on.

Comment (st3iny) [reply lib/Controller/MessagesController.php:561]: My bad. Thanks for digging the code.

Comment (ChristophWurst) [reply lib/Controller/MessagesController.php:561]: No worries. I appreciate the thorough review :pray:

**Pull request merged**